### ISSUE TITLE
Shims, backwards compatibility, transitions

### ISSUE THREAD

**Issue by rgbkrk** [jupyter/phosphor-notebook#18]
How much of the phosphor notebook will be able to handle shims, backwards compatibility, and transitions from the current JavaScript notebook APIs and namespaces?

/cc @jdfreder @Carreau

**Comment by sccolbert** [comment]
Not sure I follow. Are you asking for them to be compatible?

The most direct answer to your question is "probably 0%". When we started this, we weren't concerned with backwards compatibility, and there weren't really any namespaces in the original code. There was lots of global state to get rid of, however. It's been mentioned several times by @minrk that the frontend APIs were never listed as stable.

**Comment by rgbkrk** [comment]
As part of https://github.com/jupyter/notebook/pull/268, this was brought up about the use of requirejs, the `IPython` global object, and generally backwards compatibility. I'm raising this as an awareness since we're currently going through a transition to the `jupyter/*` python packages and should set some expectations on what would be maintained or if this is a 100% rewrite independently installed package (circumventing upgrade woes from users).

**Comment by minrk** [comment]
It depends. The original proposal was to develop the phosphor notebook only as a 5.0 that breaks all the things. We are now considering integrating at least _some_ of the pieces (e.g. services, output area) into the existing application in 4.x releases, which would need to consider what it would break, and if/how breakage might be managed. Since there are "zero stable APIs" in the javascript, I don't feel the need for perfect compatibility, but we need to be conscientious about what we break in 4.x, even on the js side.

That doesn't mean we can't do both - integrate a few pieces in 4.x releases, but still do 5.0 once bigger changes are needed, with zero compatibility.

**Comment by rgbkrk** [comment]
Yeah, I was primarily thinking about notebook extensions and libraries that rely on `IPython` in JS. This includes ipywidgets, but I imagine that's going to be one of the special cases. We have responded in mailing lists, stack overflow, and github issues about the JS API, so whether we want it or not it's the one we've been implicitly supporting and documenting.

**Comment by minrk** [comment]
I'm fine supporting a volatile API, saying "it's changed". We have historically done releases like this - run as 4.x while we can maintain backward compatibility and bump to 5.0 once we start wanting changes that aren't feasible to implement while preserving backward-compat. I'm quite sure that we will want to drop backward compat by the time we get to a full-blown refactored notebook app. It's just a question of how much we want to do before making that jump.
